#!/usr/bin/env python3
"""
Cliente de pruebas para el servidor gRPC de Messaging.

Este cliente verifica que todas las funcionalidades del servidor est√©n
funcionando correctamente, incluyendo:
- Obtenci√≥n de par√°metros de messaging
- Obtenci√≥n de routing keys
- Streaming de mensajes de schemas
- Streaming de mensajes de validation

Uso:
    python test_client.py [--host HOST] [--port PORT] [--timeout SECONDS]
"""

import argparse
import asyncio
import signal
import sys

import grpc
from messaging_utils.messaging.connection_factory import (
    RabbitMQConnectionFactory,
)
from messaging_utils.messaging.publishers import Publisher
from messaging_utils.schemas.connection import ConnectionParams
from proto_utils.generated.messaging import messaging_pb2, messaging_pb2_grpc
from proto_utils.messaging.dtypes import JsonSchema, Metadata

from src.core.config import settings


class MessagingTestClient:
    """Cliente de pruebas para el servidor gRPC de Messaging."""

    def __init__(self, host: str = "localhost", port: int = 50055):
        """
        Inicializar el cliente de pruebas.

        Args:
            host: Direcci√≥n del servidor gRPC
            port: Puerto del servidor gRPC
        """
        self.address = f"{host}:{port}"
        self.channel = None
        self.stub = None
        self.publisher = None

    async def connect(self) -> None:
        """Establecer conexi√≥n con el servidor gRPC."""
        print(f"üîå Conectando al servidor gRPC en {self.address}...")
        self.channel = grpc.aio.insecure_channel(self.address)
        self.stub = messaging_pb2_grpc.MessagingServiceStub(self.channel)

        # Verificar que el servidor est√© disponible
        try:
            await self.channel.channel_ready()
            print("‚úÖ Conexi√≥n establecida exitosamente")
        except Exception as e:
            print(f"‚ùå Error al conectar con el servidor: {e}")
            raise

    async def setup_publisher(self) -> None:
        """Configurar el publisher para enviar mensajes de prueba."""
        print("üì° Configurando publisher para mensajes de prueba...")

        try:
            # Primero obtenemos los par√°metros de messaging del servidor
            request = messaging_pb2.GetMessagingParamsRequest()
            response = await self.stub.GetMessagingParams(request)

            # Configuramos RabbitMQ con los par√°metros obtenidos
            messaging_params = {
                "host": response.host,
                "port": response.port,
                "username": response.username,
                "password": response.password,
                "virtual_host": response.virtual_host,
                "exchange": {
                    "exchange": response.exchange.exchange,
                    "type": response.exchange.type,
                    "durable": response.exchange.durable,
                    "queues": [],
                },
            }

            RabbitMQConnectionFactory.configure(messaging_params)

            # Crear publisher
            connection_params = ConnectionParams(
                host=response.host,
                port=response.port,
                virtual_host=response.virtual_host,
                username=response.username,
                password=response.password,
            )

            exchange_info = {
                "exchange": response.exchange.exchange,
                "type": response.exchange.type,
                "durable": response.exchange.durable,
            }

            self.publisher = Publisher(
                params=connection_params, exchange_info=exchange_info
            )

            print("‚úÖ Publisher configurado exitosamente")

        except Exception as e:
            print(f"‚ùå Error configurando publisher: {e}")
            raise

    async def test_get_messaging_params(self) -> None:
        """Probar la obtenci√≥n de par√°metros de messaging."""
        print("\nüß™ Prueba: GetMessagingParams")
        print("-" * 40)

        try:
            request = messaging_pb2.GetMessagingParamsRequest()
            response = await self.stub.GetMessagingParams(request)

            print(f"‚úÖ Host: {response.host}")
            print(f"‚úÖ Port: {response.port}")
            print(f"‚úÖ Virtual Host: {response.virtual_host}")
            print(f"‚úÖ Exchange: {response.exchange.exchange}")
            print(f"‚úÖ Exchange Type: {response.exchange.type}")

        except Exception as e:
            print(f"‚ùå Error: {e}")
            raise

    async def test_get_routing_keys(self) -> None:
        """Probar la obtenci√≥n de routing keys."""
        print("\nüß™ Prueba: Routing Keys")
        print("-" * 40)

        try:
            # Test schemas routing key
            schemas_request = messaging_pb2.GetRoutingKeySchemasRequest(
                results=True
            )
            schemas_response = await self.stub.GetRoutingKeySchemas(
                schemas_request
            )
            print(f"‚úÖ Schemas routing key: {schemas_response.routing_key}")

            # Test validations routing key
            validations_request = messaging_pb2.GetRoutingKeyValidationsRequest(
                results=True
            )
            validations_response = await self.stub.GetRoutingKeyValidations(
                validations_request
            )
            print(
                f"‚úÖ Validations routing key: {validations_response.routing_key}"
            )

        except Exception as e:
            print(f"‚ùå Error: {e}")
            raise

    async def test_schema_streaming(self, timeout: int = 30) -> None:
        """Probar el streaming de mensajes de schemas."""
        print(f"\nüß™ Prueba: Schema Messages Streaming (timeout: {timeout}s)")
        print("-" * 40)

        # Crear tarea para enviar mensajes de prueba
        send_task = asyncio.create_task(self._send_schema_messages())

        try:
            request = messaging_pb2.SchemaMessageRequest()

            print("üì• Iniciando streaming de mensajes de schemas...")
            messages_received = 0

            # Configurar timeout
            async def stream_with_timeout():
                async for message in self.stub.StreamSchemaMessages(request):
                    nonlocal messages_received
                    messages_received += 1
                    print(
                        f"‚úÖ Schema message #{messages_received}: {message.id}"
                    )
                    print(f"   üìÑ Import name: {message.import_name}")

                    # Para la demo, paramos despu√©s de recibir algunos mensajes
                    if messages_received >= 3:
                        break

            try:
                await asyncio.wait_for(stream_with_timeout(), timeout=timeout)
                print(
                    f"‚úÖ Streaming completado. Mensajes recibidos: {messages_received}"
                )
            except asyncio.TimeoutError:
                print(
                    f"‚è∞ Timeout despu√©s de {timeout}s. Mensajes recibidos: {messages_received}"
                )

        except Exception as e:
            print(f"‚ùå Error en streaming: {e}")
        finally:
            send_task.cancel()

    async def test_validation_streaming(self, timeout: int = 30) -> None:
        """Probar el streaming de mensajes de validation."""
        print(
            f"\nüß™ Prueba: Validation Messages Streaming (timeout: {timeout}s)"
        )
        print("-" * 40)

        # Crear tarea para enviar mensajes de prueba
        send_task = asyncio.create_task(self._send_validation_messages())

        try:
            request = messaging_pb2.ValidationMessageRequest()

            print("üì• Iniciando streaming de mensajes de validation...")
            messages_received = 0

            # Configurar timeout
            async def stream_with_timeout():
                async for message in self.stub.StreamValidationMessages(
                    request
                ):
                    nonlocal messages_received
                    messages_received += 1
                    print(
                        f"‚úÖ Validation message #{messages_received}: {message.id}"
                    )
                    print(f"   üìÑ Import name: {message.import_name}")
                    print(f"   üè∑Ô∏è Task: {message.task}")
                    print(f"   üìÅ Filename: {message.metadata.filename}")

                    # Para la demo, paramos despu√©s de recibir algunos mensajes
                    if messages_received >= 3:
                        break

            try:
                await asyncio.wait_for(stream_with_timeout(), timeout=timeout)
                print(
                    f"‚úÖ Streaming completado. Mensajes recibidos: {messages_received}"
                )
            except asyncio.TimeoutError:
                print(
                    f"‚è∞ Timeout despu√©s de {timeout}s. Mensajes recibidos: {messages_received}"
                )

        except Exception as e:
            print(f"‚ùå Error en streaming: {e}")
        finally:
            send_task.cancel()

    async def _send_schema_messages(self) -> None:
        """Enviar mensajes de schema para probar el streaming."""
        if not self.publisher:
            return

        await asyncio.sleep(2)  # Esperar a que el streaming inicie

        for i in range(5):
            try:
                test_schema = JsonSchema(
                    schema="https://json-schema.org/draft/2020-12/schema",
                    type="object",
                    properties={
                        f"field_{i}": {"type": "string"},
                        "id": {"type": "integer"},
                    },
                    required=[f"field_{i}"],
                )

                self.publisher.publish_schema_update(
                    routing_key=settings.RABBITMQ_ROUTING_KEY_SCHEMAS,
                    schema=test_schema,
                    import_name=f"test_schema_{i}",
                    raw=False,
                    task="upload_schema",
                )

                print(f"üì§ Enviado schema message #{i + 1}")
                await asyncio.sleep(2)

            except Exception as e:
                print(f"‚ùå Error enviando schema message: {e}")
                break

    async def _send_validation_messages(self) -> None:
        """Enviar mensajes de validation para probar el streaming."""
        if not self.publisher:
            return

        await asyncio.sleep(2)  # Esperar a que el streaming inicie

        for i in range(5):
            try:
                test_data = f"test data content {i}".encode()

                metadata = Metadata(
                    filename=f"test_file_{i}.txt",
                    content_type="text/plain",
                    size=len(test_data),
                )

                self.publisher.publish_validation_request(
                    routing_key=settings.RABBITMQ_ROUTING_KEY_VALIDATIONS,
                    file_data=test_data,
                    import_name=f"test_validation_{i}",
                    metadata=metadata,
                    task="sample_validation",
                )

                print(f"üì§ Enviado validation message #{i + 1}")
                await asyncio.sleep(2)

            except Exception as e:
                print(f"‚ùå Error enviando validation message: {e}")
                break

    async def run_tests(self, timeout: int = 30) -> None:
        """Ejecutar todas las pruebas."""
        print("üöÄ Iniciando pruebas del cliente gRPC de Messaging")
        print("=" * 60)

        try:
            await self.connect()
            await self.test_get_messaging_params()
            await self.setup_publisher()
            await self.test_get_routing_keys()
            await self.test_schema_streaming(timeout)
            await self.test_validation_streaming(timeout)

            print("\nüéâ Todas las pruebas completadas exitosamente!")

        except Exception as e:
            print(f"\nüí• Error durante las pruebas: {e}")
            sys.exit(1)
        finally:
            await self.close()

    async def close(self) -> None:
        """Cerrar conexiones."""
        if self.channel:
            await self.channel.close()
        if self.publisher:
            self.publisher.close()


def signal_handler(signum, frame):
    """Manejar se√±ales de interrupci√≥n."""
    print(f"\nüõë Se√±al recibida ({signum}). Cerrando cliente...")
    sys.exit(0)


async def main():
    """Funci√≥n principal."""
    parser = argparse.ArgumentParser(
        description="Cliente de pruebas para servidor gRPC de Messaging"
    )
    parser.add_argument(
        "--host",
        default="localhost",
        help="Direcci√≥n del servidor (default: localhost)",
    )
    parser.add_argument(
        "--port",
        type=int,
        default=50055,
        help="Puerto del servidor (default: 50051)",
    )
    parser.add_argument(
        "--timeout",
        type=int,
        default=30,
        help="Timeout para streaming en segundos (default: 30)",
    )

    args = parser.parse_args()

    # Configurar manejo de se√±ales
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    # Ejecutar pruebas
    client = MessagingTestClient(args.host, args.port)
    await client.run_tests(args.timeout)


if __name__ == "__main__":
    asyncio.run(main())
